src/hotspot/share/classfile/systemDictionaryShared.cpp

```diff
@@ -1537,6 +1541,76 @@ bool SystemDictionaryShared::is_dumptime_table_empty() {
   return false;
 }
 
+class CloneDumpTimeClassTable: public StackObj {
+  DumpTimeSharedClassTable* _table;
+  DumpTimeSharedClassTable* _cloned_table;
+ public:
+  CloneDumpTimeClassTable(DumpTimeSharedClassTable* table, DumpTimeSharedClassTable* clone) :
+                      _table(table), _cloned_table(clone) {
+    assert(_table != NULL, "_dumptime_table is NULL");
+    assert(_cloned_table != NULL, "_cloned_table is NULL");
+  }
+  bool do_entry(InstanceKlass* k, DumpTimeClassInfo& info) {
+    if (!info.is_excluded()) {
+      bool created;
+      _cloned_table->put_if_absent(k, info.clone(), &created);
+    }
+    return true; // keep on iterating
+  }
+};
+
+class CloneDumpTimeLambdaProxyClassTable: StackObj {
+  DumpTimeLambdaProxyClassDictionary* _table;
+  DumpTimeLambdaProxyClassDictionary* _cloned_table;
+ public:
+  CloneDumpTimeLambdaProxyClassTable(DumpTimeLambdaProxyClassDictionary* table,
+                                     DumpTimeLambdaProxyClassDictionary* clone) :
+                      _table(table), _cloned_table(clone) {
+    assert(_table != NULL, "_dumptime_table is NULL");
+    assert(_cloned_table != NULL, "_cloned_table is NULL");
+  }
+
+  bool do_entry(LambdaProxyClassKey& key, DumpTimeLambdaProxyClassInfo& info) {
+    assert_lock_strong(DumpTimeTable_lock);
+    bool created;
+    // make copies then store in _clone_table
+    LambdaProxyClassKey keyCopy = key;
+    _cloned_table->put_if_absent(keyCopy, info.clone(), &created);
+    ++ _cloned_table->_count;
+    return true; // keep on iterating
+  }
+};
+
+void SystemDictionaryShared::clone_dumptime_tables() {
+  Arguments::assert_is_dumping_archive();
+  if (_dumptime_table != NULL) {
+    assert(_cloned_dumptime_table == NULL, "_cloned_dumptime_table must be cleaned");
+    _cloned_dumptime_table = new (ResourceObj::C_HEAP, mtClass) DumpTimeSharedClassTable;
+    CloneDumpTimeClassTable copy_classes(_dumptime_table, _cloned_dumptime_table);
+    _dumptime_table->iterate(&copy_classes);
+    _cloned_dumptime_table->update_counts();
+  }
+  if (_dumptime_lambda_proxy_class_dictionary != NULL) {
+    assert(_cloned_dumptime_lambda_proxy_class_dictionary == NULL,
+           "_cloned_dumptime_lambda_proxy_class_dictionary must be cleaned");
+    _cloned_dumptime_lambda_proxy_class_dictionary =
+                                          new (ResourceObj::C_HEAP, mtClass) DumpTimeLambdaProxyClassDictionary;
+    CloneDumpTimeLambdaProxyClassTable copy_proxy_classes(_dumptime_lambda_proxy_class_dictionary,
+                                                          _cloned_dumptime_lambda_proxy_class_dictionary);
+    _dumptime_lambda_proxy_class_dictionary->iterate(&copy_proxy_classes);
+  }
+}
+
+void SystemDictionaryShared::restore_dumptime_tables() {
+  // FileMapInfo::clean_cloned_shared_path_table();
```

Comment:
Please delete line 1605 if it is not needed.

Reply:
Ah, forget to delete it. Thanks